compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs

```diff
@@ -654,6 +655,64 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
         }
     }
 
+    /// Suggests moving redundant argument(s) of an associate function to the
+    /// trait it belongs to.
+    ///
+    /// ```compile_fail
+    /// Into::into::<Option<_>>(42) // suggests considering `Into::<Option<_>>::into(42)`
+    /// ```
+    fn suggest_moving_args(&self, err: &mut Diagnostic) {
+        if let Some(trait_) = self.tcx.trait_of_item(self.def_id) {
+            // HACK(hkmatsumoto): Ugly way to tell "<trait>::<assoc fn>()" from "x.<assoc fn>()";
+            // we don't care the latter (for now).
+            if self.path_segment.res == Some(hir::def::Res::Err) {
+                return;
+            }
+
+            // Say, if the assoc fn takes `A`, `B` and `C` as generic arguments while expecting 1
+            // argument, and its trait expects 2 arguments. It is hard to "split" them right as
+            // there are too many cases to handle: `A` `B` | `C`, `A` `B` | `C`, `A` `C` | `B`, ...
+            let num_assoc_fn_expected_args =
+                self.num_expected_type_or_const_args() + self.num_expected_lifetime_args();
+            if num_assoc_fn_expected_args > 0 {
+                return;
+            }
+
+            let num_assoc_fn_excess_args =
+                self.num_excess_type_or_const_args() + self.num_excess_lifetime_args();
+
+            let trait_generics = self.tcx.generics_of(trait_);
+            let num_trait_generics_except_self =
+                trait_generics.count() - if trait_generics.has_self { 1 } else { 0 };
+
+            // FIXME(hkmatsumoto): RHS of this condition ideally should be
+            // `num_trait_generics_except_self` - "# of generic args already provided to trait"
+            // but unable to get that information with `self.def_id`.
```

Comment:
Are you able to look at the HIR at all to work out what generics were provided to the trait?

Reply:
Fixed by 2fd8577d8aeb24a9aed5a53cfd7606e4bd420342